tools/hardware/utils/memory.py

```diff
@@ -26,6 +26,36 @@ def visitor(node: WrappedNode):
     return regions
 
 
+def merge_memory_regions(regions: Set[Region]) -> Set[Region]:
+    ''' Check all region and merge adjacent ones '''
+    all_regions = [dict(idx=idx, region=region, right_adj=None, left_adj=None)
+                   for (idx, region) in enumerate(regions)]
+
+    # Find all right contiguous regions
+    for dreg in all_regions:
```

Comment:
I'm trying to understand the loging here and this seem over-complicates? Why not got over the array and build a new array with the merged region. We can modify the elements in the new array without worries.

Reply:
That's what it does, no? These are two sets, unordered. The first loop finds what is adjacent, the second builds a new set.

Doing it in one go would be fairly complex, you may have to merge one entry multiple times, and you don't want to iterate over a mutating structure.

Reply:
(Sorry, one set, not two, it just has to pair-wise compare all regions)